helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -0,0 +1,164 @@
+
+      for (Partition partition : partitionMsgs.keySet()) {
+        String partitionName = partition.getPartitionName();
+
+        // Get Partition Weight
+        Map<String, Integer> partCapacity = getPartitionCapacity(resource, partitionName);
+
+        // TODO - check
+        Map<String, Message> msgs = partitionMsgs.get(partition);
+        // TODO - Check
+        for (String instance : msgs.keySet()) {
+           reduceAvailableInstanceCapacity(instance, partCapacity);
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the partition capacity given Resource and Partition name.
+   */
+  private Map<String, Integer> getPartitionCapacity(String resource, String partition) {
+    ClusterConfig clusterConfig = _cache.getClusterConfig();
+    ResourceConfig resourceConfig = _cache.getResourceConfig(resource);
+
+
+    // Parse the entire capacityMap from ResourceConfig
+    Map<String, Map<String, Integer>> capacityMap;
+    try {
+      capacityMap = resourceConfig.getPartitionCapacityMap();
+    } catch (IOException ex) {
+      return createDefaultParticipantWeight();
+    }
+    return WagedValidationUtil.validateAndGetPartitionCapacity(partition, resourceConfig, capacityMap, clusterConfig);
+  }
+
+  /**
+   * Get the instance remaining capacity.
+   * Capacity and weight both are represented as Key-Value.
+   * Returns the capacity map of available head room for the instance.
+   * @param instanceName - instance name to query
+   * @return Map<String, Integer> - capacity pair for all defined attributes for the instance.
+   */
+  public Map<String, Integer> getInstanceAvailableCapacity(String instanceName) {
+    return _perInstance.get(instanceName);
+  }
+
+  public boolean isInstanceCapacityAvailable(String instance, Map<String, Integer> partitionCapacity) {
+    Map<String, Integer> instanceCapacity = _perInstance.get(instance);
+    for (String key : instanceCapacity.keySet()) {
+      if (partitionCapacity.containsKey(key)) {
+        Integer partCapacity = partitionCapacity.get(key);
+        if (partCapacity != 0 && instanceCapacity.get(key) < partCapacity) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public boolean reduceAvailableInstanceCapacity(String instance, Map<String, Integer> partitionCapacity) {
+    Map<String, Integer> instanceCapacity = _perInstance.get(instance);
+    boolean modified = false;
+    for (String key : instanceCapacity.keySet()) {
+      if (partitionCapacity.containsKey(key)) {
+        Integer partCapacity = partitionCapacity.get(key);
+        if (partCapacity != 0 && instanceCapacity.get(key) < partCapacity) {
+          return false;
+        }
+        if (partCapacity != 0) {
+          instanceCapacity.put(key, instanceCapacity.get(key) - partCapacity);
+          modified = true;
+        }
+      }
+    }
+    if (modified) {
+      _perInstance.put(instance, instanceCapacity);
+    }
```

Comment:
We may still have partial update with this. `instanceCapacity` is just a reference, not new copy

Reply:
addressed your comments. thanks